pkg/sql/opt/memo/statistics_builder.go

```diff
@@ -454,22 +454,18 @@ func (sb *statisticsBuilder) buildScan(scan *ScanExpr, relProps *props.Relationa
 	if scan.Constraint != nil {
 		// Calculate distinct counts for constrained columns
 		// -------------------------------------------------
-		applied := sb.applyConstraint(scan.Constraint, scan, relProps)
+		numUnappliedConjuncts := sb.applyIndexConstraint(scan.Constraint, scan, relProps)
 
 		var cols opt.ColSet
-		for i := 0; i < scan.Constraint.Columns.Count(); i++ {
+		for i := 0; i < scan.Constraint.ConstrainedColumns(sb.evalCtx); i++ {
```

Comment:
You should use the `for i, n := 0, scan.Constraint.ConstrainedColumns(sb.evalCtx)` pattern. Otherwise, you're unnecessarily recomputing the constrained column count on each iteration.